app/src/main/java/ch/epfl/sweng/swenggolf/Leaderboard.java

```diff
@@ -0,0 +1,101 @@
+package ch.epfl.sweng.swenggolf;
+
+import android.os.Bundle;
+import android.support.annotation.NonNull;
+import android.support.v7.widget.DefaultItemAnimator;
+import android.support.v7.widget.DividerItemDecoration;
+import android.support.v7.widget.LinearLayoutManager;
+import android.support.v7.widget.RecyclerView;
+import android.util.Log;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.TextView;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import ch.epfl.sweng.swenggolf.database.AttributeOrdering;
+import ch.epfl.sweng.swenggolf.database.Database;
+import ch.epfl.sweng.swenggolf.database.DbError;
+import ch.epfl.sweng.swenggolf.database.ValueListener;
+import ch.epfl.sweng.swenggolf.profile.User;
+import ch.epfl.sweng.swenggolf.tools.FragmentConverter;
+
+public class Leaderboard extends FragmentConverter {
+
+    private LeaderboardAdapter mAdapter;
+    protected RecyclerView.LayoutManager mLayoutManager;
+    private TextView errorMessage;
+    private TextView noUser;
+    public static final List<User> userList = new ArrayList<>();
+    private static final String ATTRIBUTE = "points";
+    private static final int LEADERBOARD_SIZE = 10;
+
+    @Override
+    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
+                             Bundle savedInstance) {
+        setToolbar(R.drawable.ic_menu_black_24dp, R.string.leaderboard);
+        View inflated = inflater.inflate(R.layout.activity_leaderboard, container, false);
+        errorMessage = inflated.findViewById(R.id.error_message);
+        noUser = inflated.findViewById(R.id.no_user_to_show);
+        setRecyclerView(inflated);
+        return inflated;
+    }
+
+
+    private void setRecyclerView(View inflated) {
+        noUser.setVisibility(View.VISIBLE);
+        RecyclerView mRecyclerView = inflated.findViewById(R.id.users_recycler_view);
+        mLayoutManager = new LinearLayoutManager(this.getContext());
+        mRecyclerView.setLayoutManager(mLayoutManager);
+        mRecyclerView.setItemAnimator(new DefaultItemAnimator());
+        UserListener userListener = new UserListener() {
```

Comment:
Since you use this listener only once here, wouldn't it be better to pass the fragment to the adapter constructor and then just do the method call inside the onClick at line 98 in the adapter (thus deleting the UserListener, avoiding nested listeners and passing it as an argument) ? Plus you could then remove the context as an argument of the adapter and get the context from the fragment.

Reply:
I can't see how we can do it!

Reply:
It's pretty easy : 
1. change the adapter constructor to take 1. the list (just as before) 2. a FragmentConverter
2. on creation of the adapter pass "this" as an argument
3. you still might have your attribute "mContext" but initialized to (where mFragment is the FragmentConverter passed as an argument in the constructor) mFragment.getContext()
4. in your listener just replace your call to the UserListener by "mFragment.replaceCentralFragment(...)"

Reply:
Oh thanks it's really better now!